expr/func_highestlowest.go

```diff
@@ -0,0 +1,69 @@
+package expr
+
+import (
+	"sort"
+
+	"github.com/grafana/metrictank/consolidation"
+
+	"github.com/grafana/metrictank/api/models"
+)
+
+type FuncHighestLowest struct {
+	in      GraphiteFunc
+	n       int64
+	fn      string
+	highest bool
+}
+
+func NewHighestLowestConstructor(fn string, highest bool) func() GraphiteFunc {
+	return func() GraphiteFunc {
+		return &FuncHighestLowest{fn: fn, highest: highest}
+	}
+}
+
+func (s *FuncHighestLowest) Signature() ([]Arg, []Arg) {
+	if s.fn != "" {
+		return []Arg{
+			ArgSeriesList{val: &s.in},
+			ArgInt{key: "n", val: &s.n},
+		}, []Arg{ArgSeriesList{}}
+	}
+	return []Arg{
+		ArgSeriesList{val: &s.in},
+		ArgInt{key: "n", val: &s.n},
+		ArgString{key: "func", val: &s.fn, validator: []Validator{IsConsolFunc}},
+	}, []Arg{ArgSeriesList{}}
+}
+
+func (s *FuncHighestLowest) Context(context Context) Context {
+	return context
+}
+
+func (s *FuncHighestLowest) Exec(cache map[Req][]models.Series) ([]models.Series, error) {
+	series, err := s.in.Exec(cache)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(series) == 0 {
+		return series, nil
+	}
+
+	consolidationFunc := consolidation.GetAggFunc(consolidation.FromConsolidateBy(s.fn))
+
+	seriesLess := func(i, j int) bool {
+		si := consolidationFunc(series[i].Datapoints)
+		sj := consolidationFunc(series[j].Datapoints)
```

Comment:
You will likely end up running `consolidationFunc` on the same series many times by doing it in the function. Perhaps a struct that contains a pointer and a value to cache the result of the consolidation? I'd be interested in the benchmark performance difference

Reply:
yeah I think it uses quicksort so about nlog(n) times. Shouldn't be too much of a performance issue, but I'll add in a struct

Reply:
just added a slice